### ISSUE TITLE
TensorRT server  stucked when I run perf_client command 

### ISSUE THREAD

**Issue by diligent321** [NVIDIA/tensorrt-inference-server#456]
hello, I have trained an inception_resnet_v2 model(.pb), about 220MB, then i use function create_inference_graph() to convert to tf-trt FP32(bigger than 400MB), tf-trt FP16(bigger than 300MB),

then , I make pressure test with perf_client command, the result is,

when the server is runing a tf model, the return statics is normal. while when the server is running a tf-trt FP32 model, only the first GPU memory rise until occupy all memory, then the server stucked.

so what's the problem?

by the way, although I set CUDA_VISIBLE_DEVICES=0, when I run the tensorrt server, it occupy all GPUs, how to assign tensorrt server runing on some GPU?

my hardware is T4 with 4 cards.

**Comment by bezero** [comment]
@diligent321 You should probably look at your max batch size. The sizes for the converted models seems to be correct. When you use tensorrt, it has fixed size that it will occupy on GPU memory and it depends on your max batch size. You might have specified too big value for max batch. In TF case, GPU memory is utilized during run-time based on incomming request. Therefore, you might have observed this discrepancy for your tf vs tf-trt models. How are you setting CUDA_VISIBLE_DEVICES? Normally, I would set it when I run trtis docker container by passing env variable as:
 ```
$ docker run --rm --shm-size=1g --ulimit memlock=-1 --ulimit stack=67108864 -e CUDA_VISIBLE_DEVICES=0 -p8000:8000 -p8001:8001 -p8002:8002 -v/path/to/model/repository:/models <tensorrtserver image name> trtserver --model-store=/models
```

**Comment by diligent321** [comment]
Thans for your kindly reply. The out of memory error is because of the large batch size. now, it can run normally.  when CUDA_VISIBLE_DEVICES=0 is included in docker command, it occcupy only one gpu.

Now, i have another question, i run model in all 4 cards, when i make pressure test with perf_client, it seems only the first card runs, as the following image shows,
![捕获](https://user-images.githubusercontent.com/31379237/61195148-eb391e00-a6f8-11e9-9392-44526f9f770d.PNG)
and the return statics are almost the same as running on one card, so how to distribute http posts on all 4 cards ? @bezero

**Comment by bezero** [comment]
@diligent321 With the above configuration your model will be running only on 1st GPU. You can add more GPU by ```CUDA_VISIBLE_DEVICES=0,1,2``` for example. If you remove ```CUDA_VISIBLE_DEVICES``` it will use all GPU resources available. If you want to allocate all the available resources to TRTIS, but you want to limit your deployed model to be running on certain GPUs, you can set this option on your models' deployment ```congig.pbtxt``` by setting ```instance_group```. For example:
```
instance_group [
  {
    count: 1
    kind: KIND_GPU
    gpus: [ 0, 1, 2 ]
  }
]
```
To be able to utilise all your resources, you could send multiple async requests, so that your data are being processed in parallel by different instances. As far as I know, you can not distribute single request execution to multiple GPU instances, but you have to verify this hypothesis.

**Comment by diligent321** [comment]
@bezero thanks, maybe i did not express  my question clearly, i clarify it here again, 
In my server machine, i run the following script, 

nvidia-docker run --rm --shm-size=1g --ulimit memlock=-1 --ulimit stack=67108864           -e CUDA_VISIBLE_DEVICES=0,1,2,3  -p8000:8000 -p8001:8001 -p8002:8002 -v/model_dir:/models d98ba9ab7916 trtserver --model-store=/models

the instance_group in config.pbtxt is as follows,

 20 instance_group [
 21   {
 22     kind: KIND_GPU,
 23     count: 4
 24   }
 25 ]

when i start the server, the GPU usage is as follows, 
![捕获](https://user-images.githubusercontent.com/31379237/61202234-91951b80-a719-11e9-8021-177f62e2e4c9.PNG)


In my client machine, i run the following script,

 perf_client -m  inception_resnet_v2_graphdef  -p30000 -t16 -v -u 10.102.150.145:8000

 the GPU usage is as follows, 
![捕获](https://user-images.githubusercontent.com/31379237/61202400-010b0b00-a71a-11e9-9327-c246e79f9d28.PNG)

in my opinion, the server is running normally on 4 cards, with 4 instances on each card, but why only the first card processes inference computation when i send https with 16 concurrencies?

**Comment by bezero** [comment]
@diligent321 Based on the Options available for [perf_client](https://github.com/NVIDIA/tensorrt-inference-server/blob/d3f545d85ba088cb9ffc795f8a95355b156408d8/src/clients/c%2B%2B/perf_client.cc)
```
Options:
 -b: batch size for each request sent.
 -t: number of concurrent requests sent. If -d is set, -t indicate the number
     of concurrent requests to start with ("starting concurrency" level).
 -d: enable dynamic concurrent request mode.
 -l: latency threshold in msec, will have no effect if -d is not set.
 -p: time interval for each measurement window in msec.
```
I think, you should set concurrent request mode and set the number of concurrent requests. It should be something like this:
```
perf_client -d -t 4 -m  inception_resnet_v2_graphdef  -p30000 -t16 -v -u 10.102.150.145:8000
```

However, I am not sure. You can try running with these options.

**Comment by diligent321** [comment]
I tried this ,but with no luck, thanks all the same.  @bezero

**Comment by deadeyegoodwin** [comment]
Using perf_client -t16 should cause 16 concurrent requests to be sent to the server, and because you have 16 model instances across 4 GPUs you should see all 4 GPUs busy. What version of TRTIS are you using? Also, can you add "-f <file>" switch to your perf_client run and share the resulting CSV file.

**Comment by diligent321** [comment]
@deadeyegoodwin 
TRTIS  version:  I pulled the docker image nvcr.io/nvidia/tensorrtserver:19.05-py3.

In the console log of the server, it prints log like the following ,
I0716 01:37:20.530781 1 base_bundle.cc:162] Creating instance inception_resnet_v2_graphdef_0_{i}_gpu{j} on GPU {j} (7.5) using model.graphdef"
i={0, 1, 2, 3}, j={0, 1, 2, 3}
obviously, it create 4 model instances on each of the 4GPUs.

In the client, i run the following script, but no resulting CSV file exists, 
```
perf_client -m  inception_resnet_v2_trt_FP32_zhipeng_graphdef  -ftemp_result.csv  -p30000 -t16 -v -u 10.102.150.145:8000
```
or 
```
perf_client -m  inception_resnet_v2_trt_FP32_zhipeng_graphdef  -f  temp_result.csv  -p30000 -t16 -v -u 10.102.150.145:8000
```
the log in the client console is as follows,
![捕获](https://user-images.githubusercontent.com/31379237/61260780-221f3a80-a7b2-11e9-8f76-b36e4af66e30.PNG)

**Comment by deadeyegoodwin** [comment]
Thanks for the report. We'll try to reproduce on our end. It is strange that the CSV file is not created...

**Comment by diligent321** [comment]
Thanks for your kindly help, i do experiment again, the result is the same. When running a tf model, the model instances can run on several gpus normally, however, when running a tf-trt model, just convert the tf model to a FP32/FP16 one, it only runs on the first gpu, even if i assign "gpus: [ 2, 3 ]" in the config.pbtxt.

Really appreciate your suggestion.

**Comment by deadeyegoodwin** [comment]
This is likely related to the same issue as in #147 (but no caffe2 specific). Tensorflow API/config has a property that is supposed to allow you to control which GPUs a model runs on, but it doesn't work. As a result TRTIS is forced to use a work-around which is likely not working correctly in all cases. It could be that the TF-TRT model is explicitly placing every node on device 0. The TRTIS work-around would likely not override this explicit placement and so the model will always run on 0 (but this is just a guess we will have to verify).

**Comment by diligent321** [comment]
Different from object detection algorithm involving CPU and GPU ops, the model is inception_resnet_v2， simply a classfication model, which could be run on only GPUs, 

so I think it is a bug of TRTIS, and also look forward to your verification.

**Comment by diligent321** [comment]
@deadeyegoodwin 

Hi,
the TRTIS has many advantages, like multiple framework support, dynamic batch and so on. I would like to replace my flask framework with TRTIS as soon as possible. However, TF-TRT model can only be run on the first GPU, which limit the use in product environment.

so, is there a way to let TR-TRT models run normally on multi GPUs, instead of only the first GPU. if you have any suggestions about what I should do to deal with this disadvantage?

Sincerely look forward to your answer.

**Comment by tanmayv25** [comment]
@diligent321 

Hi,
In our verification we found TF-TRT models specifies device field in GraphDef to a single GPU. TRTIS does sets a default gpu device id, but unfortunately this setting doesn't override the specification in graphdef. We have a workaround in progress to circumvent this issue. In the mean time you can try clearing this field in the graphdef and then try running trtserver with your modified graphdef.

However, ensure the model of gpus is identical to what the model was generated on. TF-TRT models are very specific to the model they were generated on.

**Comment by diligent321** [comment]
Thanks very much,  i will try  this  method.


发自我的iPhone

------------------ Original ------------------

**Comment by diligent321** [comment]
@tanmayv25 

It is true that TF-TRT models specifies device field in GraphDef to a single GPU, even if i set CUDA_VISIBLE_DEVICES=0,1,2, the GraphDef is shown as follows, 
![捕获](https://user-images.githubusercontent.com/31379237/62030683-32a2cc80-b218-11e9-805b-73cb5f5b08da.PNG)

so, how to clear device field in the TF-TRT model, could you offer the tool script, or explain this in more detail？

**Comment by tanmayv25** [comment]
@diligent321 ,

Please have a look at the PR "Overriding the gpu allocation in GraphDef models #507" which solves your problem. The feature will soon be available at the master.

**Comment by diligent321** [comment]
Great, thanks very much， look forward to the new tensorrt server release.

**Comment by tanmayv25** [comment]
@diligent321,
The PR "Overriding the gpu allocation in GraphDef models #507" has been merged into master. The fix will be available in 19.08 release. I am closing the issue.
